debugging-modules/README.md

```diff
@@ -0,0 +1,264 @@
+      list.
+
+When a user selects a source from the list of all source files, a debugger GUI
+will typically display the selected source's text in a tab or panel. The text
+for a source can be retrieved by calling `SourceDebugger.getSourceText`.
+
+### Setting a Breakpoint
+
+When a user asks the debugger to set a source-level breakpoint, the debugger
+should perform the following steps:
+
+* Get the `SourceDebugger` associated with the breakpoint's source.
+* Let `options` be the `SourceBreakpointOptions` describing the requested
+  breakpoint.
+* Let `id` be the result of calling `SourceDebugger.setBreakpoint` with
+  `options`.
+    * If there is no expression at the breakpoint's location:
+        *  The `SourceDebugger.setBreakpoint` implementation should return null.
+    * Otherwise:
+        * The `SourceDebugger` implementation should set as many Wasm
+          breakpoints as necessary to pause just before evaluation of the
+          breakpoint location's expression begins using
+          `WasmDebugger.setBreakpoint`.
+* If `id` is null:
+    * Then no breakpoint was set (e.g because the requested location is within a
+      comment) and the UI should reflect this.
+* Otherwise:
+    * Save the `id` for if/when the breakpoint gets hit or the user wants to
+      clear the breakpoint.
+
+### Hitting a Breakpoint
+
+* Let `dbg` be the `SourceDebugger` instance for the current frame.
+* Construct the `WasmBreakInfo` for the Wasm breakpoint that was hit.
+* Let `breakResult` be the result of calling `dbg.onBreak` with the
+  `WasmBreakInfo`
+* If `breakResult.kind` is `"Continue"`:
+    * Resume the debuggee module's execution.
+* Otherwise if `breakResult.kind` is `"Pause"`:
+    * Assert that `breakResult.location` is not null.
+    * Update the debugger UI to show that execution is paused at the source
+      location `breakResult.location`.
+
+### Stepping
+
+* Let `dbg` be the `SourceDebugger` instance for the current frame.
+* Construct the `SourceStepOptions` for the step that the user requested.
+* Call `dbg.onStep` with the `SourceStepOptions`
+    * The `onStep` implementation should use the `WasmDebugger` to set Wasm
+      breakpoint(s) where it determines execution should pause after taking the
+      requested step.
+* Resume debuggee execution.
+
+Pausing after the step has been executed is equivalent to hitting a breakpoint,
+as described above, although additionally `SourceDebugger` implementations
+should delete any Wasm breakpoints that were created for the step with
+`WasmDebugger.clearBreakpoint`.
+
+## Interface Definitions
+
+> **Note:** In order to be embedder agnostic — about whether it is a Web browser
+> or JavaScript engine or not — we intend to eventually define these interfaces
+> with [WebAssembly Interface
+> Types](https://github.com/WebAssembly/interface-types/). However, since that
+> standard is still coming together, we are temporarily describing the
+> interfaces with Web IDL.
```

Comment:
It could be more clearly noted earlier in the document that this is not meant to be specific to an embedding that uses JS, as that is the impression I got from this and earlier documents until I saw this.